### PR TITLE
feat: add killswitch for notifications

### PR DESCRIPTION
**Pull request by mtojek**
Related: https://github.com/coder/team-coconut/issues/4
Blocked: https://github.com/coder/coder/pull/13537

This PR introduces API endpoints to pause/unpause the notifier. The setting is stored in the database. In the follow up I will add CLI support to control the setting.

### CODE REVIEW

File: coderd/database/dbpurge/dbpurge_test.go [hunk 1]
```diff
@@ -205,8 +205,8 @@ func TestDeleteOldWorkspaceAgentLogs(t *testing.T) {
 			}
 			return !containsAgentLog(agentLogs, t.Name())
 		}, testutil.WaitShort, testutil.IntervalFast)
-		require.NoError(t, err)
-		require.NotContains(t, agentLogs, t.Name())
+		assert.NoError(t, err)
```
Comment (mtojek) [inline coderd/database/dbpurge/dbpurge_test.go:208]: drive-by change: I'm fighting [this flake](https://github.com/coder/coder/issues/13801).

File: coderd/notifications/notifications_test.go [hunk 2]
```diff
@@ -517,6 +517,113 @@ func TestInvalidConfig(t *testing.T) {
...
+	settingsJSON, err = json.Marshal(&codersdk.NotificationsSettings{
+		NotifierPaused: false,
+	})
+	require.NoError(t, err)
+	err = db.UpsertNotificationsSettings(ctx, string(settingsJSON))
+	require.NoError(t, err)
+
+	// test
+	// Check if messages have been dispatched
+	require.Eventually(t, func() bool {
```
Comment (dannykopping) [inline coderd/notifications/notifications_test.go:615]: Nit: I'd suggest looking at the `TestBasicNotificationRoundtrip` which uses `fakeHandler` or `TestBufferedUpdates` which uses the `bulkUpdateInterceptor`. You don't need postgres or the webhook instrumentation for this test.

It'd also feel more realistic to have the notifier process a message first, then pause, then validate paused, then unpause, then validate it works as expected.

File: coderd/notifications_test.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+			NotifierPaused: true,
+		}
+
+		ctx := testutil.Context(t, testutil.WaitShort)
+
+		// when
+		err := anotherClient.PutNotificationsSettings(ctx, expected)
+
+		// then
+		require.Error(t, err) // Insufficient permissions to update notifications settings.
```
Comment (dannykopping) [inline coderd/notifications_test.go:34]: Let's check for the specific error we're expecting to receive to prevent false-positives.

File: coderd/notifications_test.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+		// given
+		expected := codersdk.NotificationsSettings{
+			NotifierPaused: false,
+		}
+		err := client.PutNotificationsSettings(ctx, expected)
+		require.NoError(t, err)
+
+		// then
+		err = client.PutNotificationsSettings(ctx, expected)
+		require.Error(t, err) // Error: notifications settings not modified
```
Comment (dannykopping) [inline coderd/notifications_test.go:77]: Ditto

Comment (dannykopping) [reply coderd/notifications_test.go:77]: Why should another call to `PutNotificationsSettings` with the same contents return an error?

**Review by dannykopping — COMMENTED** [review]
Overall looks good! Mostly small comments and a few nits.

File: coderd/database/dbpurge/dbpurge_test.go (hunk 1 above)
Comment (mtojek) [reply coderd/database/dbpurge/dbpurge_test.go:208]: It does not address, but should give more insights into real values of `agentLogs` and `err`. I can revert it though.

File: coderd/database/dbauthz/dbauthz.go [hunk 5]
```diff
@@ -1479,6 +1479,11 @@ func (q *querier) GetNotificationMessagesByStatus(ctx context.Context, arg datab
 	return q.db.GetNotificationMessagesByStatus(ctx, arg)
 }
 
+func (q *querier) GetNotificationsSettings(ctx context.Context) (string, error) {
+	// No authz checks
```
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/database/dbauthz/dbauthz.go:1483]: Actually, we expose it behind the API endpoint, so it should be accessible to at least admin users (+system).

File: coderd/notifications/notifications_test.go [hunk 6]
```diff
@@ -517,6 +517,113 @@ func TestInvalidConfig(t *testing.T) {
 	require.ErrorIs(t, err, notifications.ErrInvalidDispatchTimeout)
 }
 
+func TestNotifierPaused(t *testing.T) {
+	t.Parallel()
+
+	if !dbtestutil.WillUsePostgres() {
```
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/notifications/notifications_test.go:523]: Consistency with other test functions in the unit. Do you think we should reverify if all of them need PostgreSQL?

[the remark this answers is not in the document]
Comment (dannykopping) [reply coderd/notifications/notifications_test.go:523]: Only the tests that need Postgres have this, I checked the tests recently during Spike's review.

[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/notifications/notifications_test.go:523]: curious: if I remove `if !dbtestutil.WillUsePostgres() { ... }` from `TestWebhookDispatch`, it works on my mac?

[the remark this answers is not in the document]
Comment (dannykopping) [reply coderd/notifications/notifications_test.go:523]: Hhmm, I guess I didn't check closely enough. Thanks, I'll go through them again 👍

File: coderd/database/dbpurge/dbpurge_test.go (hunk 1 above)
Comment (mtojek) [reply coderd/database/dbpurge/dbpurge_test.go]: reverted

File: coderd/database/migrations/000223_notifications_settings_audit.down.sql [hunk 7]
```diff
@@ -0,0 +1 @@
+-- Nothing to do
```
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/database/migrations/000223_notifications_settings_audit.down.sql:1]: Fixed

File: coderd/notifications.go [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+			Detail:  err.Error(),
+		})
+		return
+	}
+
+	currentSettingsJSON, err := api.Database.GetNotificationsSettings(r.Context())
+	if err != nil {
+		httpapi.Write(r.Context(), rw, http.StatusInternalServerError, codersdk.Response{
+			Message: "Failed to fetch current notification settings.",
+			Detail:  err.Error(),
```
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/notifications.go:86]: I agree, although this is a consistent approach with other APIs, so I'd rather leave it here and have it fixed globally.

File: coderd/notifications.go [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+		})
+		return
+	}
+
+	if bytes.Equal(settingsJSON, []byte(currentSettingsJSON)) {
+		// See: https://www.rfc-editor.org/rfc/rfc7231#section-6.3.5
```
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/notifications.go]: Fixed, thanks

File: coderd/notifications.go [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+	aReq.New = database.NotificationsSettings{
+		ID:             uuid.New(),
+		NotifierPaused: settings.NotifierPaused,
+	}
+
+	err = api.Database.UpsertNotificationsSettings(ctx, string(settingsJSON))
+	if err != nil {
+		httpapi.Write(r.Context(), rw, http.StatusInternalServerError, codersdk.Response{
+			Message: "Failed to update notifications settings.",
```
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/notifications.go:114]: Fixed, rephrased to "notifications settings" everywhere.

File: coderd/notifications/notifier.go [hunk 11]
```diff
@@ -89,6 +97,25 @@ func (n *notifier) run(ctx context.Context, success chan<- dispatchResult, failu
 	}
 }
 
+// ensureRunning checks if notifier is not paused.
+func (n *notifier) ensureRunning(ctx context.Context) (bool, error) {
+	n.log.Debug(ctx, "check if notifier is paused")
```
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/notifications/notifier.go:102]: added an extra log message to the function `notifier is paused, notifications will not be delivered`

File: coderd/notifications/notifier.go [hunk 12]
```diff
@@ -71,10 +71,18 @@ func (n *notifier) run(ctx context.Context, success chan<- dispatchResult, failu
 		default:
 		}
 
-		// Call process() immediately (i.e. don't wait an initial tick).
-		err := n.process(ctx, success, failure)
+		// Check if notifier is not paused.
+		ok, err := n.ensureRunning(ctx)
 		if err != nil {
-			n.log.Error(ctx, "failed to process messages", slog.Error(err))
+			n.log.Error(ctx, "failed to check notifier state", slog.Error(err))
```
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/notifications/notifier.go:77]: > Maybe it should be a Warn since it's not breaking anything

Good call 👍 

> I think we should also log when the notifier is paused.

Done

File: coderd/notifications/notifier.go [hunk 13]
```diff
@@ -71,10 +71,18 @@ func (n *notifier) run(ctx context.Context, success chan<- dispatchResult, failu
...
+		}
+
+		if ok {
```
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/notifications/notifier.go:80]: I agree about the leftmost convention, but a simple `continue` may not be possible without an extra wrapping layer as we need to execute this logic in the end: `// Shortcut to bail out quickly...`.

Do you have a preferrence how we can wrap it?

File: coderd/notifications/notifier.go [hunk 14]
```diff
@@ -89,6 +97,25 @@ func (n *notifier) run(ctx context.Context, success chan<- dispatchResult, failu
...
+func (n *notifier) ensureRunning(ctx context.Context) (bool, error) {
+	n.log.Debug(ctx, "check if notifier is paused")
+
+	settingsJSON, err := n.store.GetNotificationsSettings(ctx)
+	if err != nil {
+		return false, xerrors.Errorf("get notifications settings: %w", err)
+	}
+
+	var settings codersdk.NotificationsSettings
+	if len(settingsJSON) > 0 {
```
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/notifications/notifier.go]: Done
BTW We should `return true` as it is unpaused 👍

File: coderd/notifications/notifications_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/notifications/notifications_test.go]: Done

File: coderd/notifications.go [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+		return
+	}
+
+	if bytes.Equal(settingsJSON, []byte(currentSettingsJSON)) {
+		// See: https://www.rfc-editor.org/rfc/rfc7231#section-6.3.5
+		httpapi.Write(r.Context(), rw, http.StatusNoContent, nil)
+		return
+	}
+
+	auditor := api.Auditor.Load()
```
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/notifications.go:97]: Hm, we do not log anywhere else? This would intrigue operators why is verbose in notifications.

File: coderd/notifications/notifications_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/notifications/notifications_test.go]: It looks we need to limit it to PostgreSQL? see below:

```
=== FAIL: coderd/notifications TestNotifierPaused (0.64s)
    notifications_test.go:520: 
        	Error Trace:	C:/a/coder/coder/coderd/notifications/utils_test.go:27
        	            				C:/a/coder/coder/coderd/notifications/notifications_test.go:520
        	Error:      	Received unexpected error:
        	            	could not start resource:
        	            	    github.com/coder/coder/v2/coderd/database/dbtestutil.OpenContainerized
        	            	        C:/a/coder/coder/coderd/database/dbtestutil/postgres.go:113
        	            	  - image operating system "linux" cannot be used on this platform: operating system is not supported
        	Test:       	TestNotifierPaused
```

```
=== FAIL: coderd/notifications TestNotifierPaused (0.00s)
    notifications_test.go:520: 
        	Error Trace:	/Users/runner/work/coder/coder/coderd/notifications/utils_test.go:27
        	            				/Users/runner/work/coder/coder/coderd/notifications/notifications_test.go:520
        	Error:      	Received unexpected error:
        	            	could not start resource:
        	            	    github.com/coder/coder/v2/coderd/database/dbtestutil.OpenContainerized
        	            	        /Users/runner/work/coder/coder/coderd/database/dbtestutil/postgres.go:113
        	            	  - dial unix /var/run/docker.sock: connect: no such file or directory
        	Test:       	TestNotifierPaused
```

File: coderd/notifications_test.go (hunk 4 above)
Comment (mtojek) [reply coderd/notifications_test.go:77]: This way we can indicate that it was HTTP 304 (see below). It is a common pattern in Coder client?

```
	if res.StatusCode == http.StatusNotModified {
		return xerrors.New("notifications settings not modified")
	}
```

File: coderd/notifications/notifications_test.go [hunk 16]
```diff
@@ -517,6 +517,113 @@ func TestInvalidConfig(t *testing.T) {
...
+	cfg.FetchInterval = *serpent.DurationOf(&fetchInterval)
+	mgr, err := notifications.NewManager(cfg, db, logger.Named("manager"))
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		assert.NoError(t, mgr.Stop(ctx))
+	})
+	enq, err := notifications.NewStoreEnqueuer(cfg, db, defaultHelpers(), logger.Named("enqueuer"))
+	require.NoError(t, err)
+
+	user := dbgen.User(t, db, database.User{
```
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/notifications/notifications_test.go]: Fixed

File: coderd/notifications/notifications_test.go [hunk 17]
```diff
@@ -517,6 +517,113 @@ func TestInvalidConfig(t *testing.T) {
...
+
+	// Start notification manager
+	mgr.Run(ctx)
+
+	// when
+	// Enqueue a bunch of messages
+	messagesCount := 50
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
```
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/notifications/notifications_test.go:566]: My intention was to send events asynchronously, and in the meantime check the queue length with `GetNotificationMessagesByStatus`, but I agree it is not superb. I will refactor this test to what you suggested below.

File: coderd/notifications_test.go (hunk 3 above)
Comment (mtojek) [reply coderd/notifications_test.go]: Fixed

File: coderd/notifications.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (dannykopping) [reply coderd/notifications.go:97]: OK fair enough 👍

File: coderd/notifications_test.go (hunk 4 above)
Comment (mtojek) [reply coderd/notifications_test.go:81]: > Even so, we should check for the very specific outcome we want, not just for any error.

Ok, I will just return `nil` and won't indicate an error

Comment (mtojek) [reply coderd/notifications_test.go]: Fixed

File: coderd/notifications_test.go [hunk 18]
```diff
@@ -60,24 +60,4 @@ func TestUpdateNotificationsSettings(t *testing.T) {
 		require.NoError(t, err)
 		require.Equal(t, expected, actual)
 	})
-
-	t.Run("Settings not modified", func(t *testing.T) {
```
Comment (dannykopping) [inline coderd/notifications_test.go]: This seems like a useful test; curious why you removed it?

Comment (mtojek) [reply coderd/notifications_test.go]: It just tests that `PutNotificationsSettings` does not fail, so I wasn't convinced about its value. I can bring it back, not a big deal 👍

File: coderd/notifications.go [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+// @Summary Update notifications settings
+// @ID update-notifications-settings
+// @Security CoderSessionToken
+// @Accept json
+// @Produce json
+// @Tags General
+// @Param request body codersdk.NotificationsSettings true "Notification settings request"
+// @Success 200 {object} codersdk.NotificationsSettings
```
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/notifications.go:56]: I added an entry for HTTP 304 to swagger annotations.

File: coderd/notifications_test.go [hunk 20]
```diff
@@ -60,4 +60,22 @@ func TestUpdateNotificationsSettings(t *testing.T) {
...
+		ctx := testutil.Context(t, testutil.WaitShort)
+
+		// given
+		expected := codersdk.NotificationsSettings{
+			NotifierPaused: false,
+		}
+
+		// then
+		err := client.PutNotificationsSettings(ctx, expected)
+		require.NoError(t, err)
```
Comment (dannykopping) [inline coderd/notifications_test.go:79]: Sorry to keep riding this test.

The absence of an error isn't specific enough for this test, I feel.
To validate that this endpoint is idempotent, we should:

1. start from an empty state
2. change the state
3. validate the state
4. try change the state again
5. validate the state is unchanged from step 3

Comment (mtojek) [reply coderd/notifications_test.go:92]: Refactored 👍 Let me know your thoughts.

Comment (dannykopping) [reply coderd/notifications_test.go:92]: Perfect!

File: coderd/notifications/notifications_test.go (hunk 2 above)
Comment (mtojek) [reply coderd/notifications/notifications_test.go:583]: Refactored 👍  Feel free to take a look.

File: coderd/notifications/notifications_test.go [hunk 21]
```diff
@@ -517,6 +517,76 @@ func TestInvalidConfig(t *testing.T) {
...
+	require.NoError(t, err)
+
+	// Notifier is paused, enqueue the next message.
+	sid, err = enq.Enqueue(ctx, user.ID, notifications.TemplateWorkspaceDeleted, map[string]string{"type": "success"}, "test")
+	require.NoError(t, err)
+	require.Eventually(t, func() bool {
+		pendingMessages, err := db.GetNotificationMessagesByStatus(ctx, database.GetNotificationMessagesByStatusParams{
+			Status: database.NotificationMessageStatusPending,
+		})
+		if err != nil {
```
Comment (dannykopping) [inline coderd/notifications/notifications_test.go:570]: I think this should perhaps be an `assert.NoError(t, err)` maybe?

File: coderd/notifications/notifications_test.go [hunk 22]
```diff
@@ -517,6 +517,76 @@ func TestInvalidConfig(t *testing.T) {
...
+	}, testutil.WaitShort, testutil.IntervalFast)
+
+	// Unpause the notifier.
+	settingsJSON, err = json.Marshal(&codersdk.NotificationsSettings{NotifierPaused: false})
+	require.NoError(t, err)
+	err = db.UpsertNotificationsSettings(ctx, string(settingsJSON))
+	require.NoError(t, err)
+
+	// Notifier is running again, message should be dequeued.
+	require.Eventually(t, func() bool {
```
Comment (dannykopping) [inline coderd/notifications/notifications_test.go:583]: Sweet, this test looks really comprehensive now 💪

**Review by dannykopping — APPROVED** [review]
LGTM! Last couple nits, none of them blocking

File: coderd/notifications/notifications_test.go [hunk 23]
```diff
@@ -520,98 +520,70 @@ func TestInvalidConfig(t *testing.T) {
...
-	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
-		w.WriteHeader(http.StatusNoContent)
-	}))
-	defer server.Close()
-
-	endpoint, err := url.Parse(server.URL)
-	require.NoError(t, err)
+	// setup
+	ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitLong)
+	t.Cleanup(cancel)
```
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/notifications/notifications_test.go]: Fixed!

File: coderd/notifications/notifications_test.go (hunk 21 above)
Comment (mtojek) [reply coderd/notifications/notifications_test.go]: Good idea 👍

**Comment by mtojek** [comment]
@dannykopping It looks like there is some flakiness around `bulkUpdate`? https://github.com/coder/coder/actions/runs/9874898408/job/27270328403?pr=13794. Not sure what is the root cause of messages not being dequeued.

```
    t.go:99: 2024-07-10 13:15:17.886 [debu]  manager.notifier: dispatch completed  notifier_id=97bfe604-a0be-4caf-b276-db37b98e3a26  count=1
    t.go:99: 2024-07-10 13:15:17.887 [debu]  manager.notifier: check if notifier is paused  notifier_id=97bfe604-a0be-4caf-b276-db37b98e3a26
    t.go:99: 2024-07-10 13:15:17.887 [debu]  manager.notifier: attempting to dequeue messages  notifier_id=97bfe604-a0be-4caf-b276-db37b98e3a26
    t.go:99: 2024-07-10 13:15:17.887 [debu]  manager.notifier: dequeued messages  notifier_id=97bfe604-a0be-4caf-b276-db37b98e3a26  count=1
    t.go:99: 2024-07-10 13:15:17.887 [debu]  manager.notifier: message dispatch succeeded  notifier_id=97bfe604-a0be-4caf-b276-db37b98e3a26  msg_id=4681beed-a763-41a1-badc-b205bfe03ed8  method=smtp
    t.go:99: 2024-07-10 13:15:17.887 [debu]  manager.notifier: dispatch completed  notifier_id=97bfe604-a0be-4caf-b276-db37b98e3a26  count=1
    t.go:99: 2024-07-10 13:15:17.887 [debu]  manager.notifier: check if notifier is paused  notifier_id=97bfe604-a0be-4caf-b276-db37b98e3a26
    t.go:99: 2024-07-10 13:15:17.887 [debu]  manager.notifier: attempting to dequeue messages  notifier_id=97bfe604-a0be-4caf-b276-db37b98e3a26
    t.go:99: 2024-07-10 13:15:17.887 [debu]  manager.notifier: dequeued messages  notifier_id=97bfe604-a0be-4caf-b276-db37b98e3a26  count=1
    t.go:99: 2024-07-10 13:15:17.887 [debu]  manager.notifier: message dispatch succeeded  notifier_id=97bfe604-a0be-4caf-b276-db37b98e3a26  msg_id=4681beed-a763-41a1-badc-b205bfe03ed8  method=smtp
    t.go:99: 2024-07-10 13:15:17.906 [debu]  enqueuer: enqueued notification  msg_id=74f1fd65-d602-42e4-87c9-a83a91abba9f
    t.go:99: 2024-07-10 13:15:19.881 [debu]  manager: bulk update completed  type=update_sent  updated=-1
    t.go:99: 2024-07-10 13:15:19.881 [debu]  manager.notifier: dispatch completed  notifier_id=97bfe604-a0be-4caf-b276-db37b98e3a26  count=1
    t.go:99: 2024-07-10 13:15:19.882 [debu]  manager.notifier: check if notifier is paused  notifier_id=97bfe604-a0be-4caf-b276-db37b98e3a26
    t.go:99: 2024-07-10 13:15:19.882 [debu]  manager.notifier: attempting to dequeue messages  notifier_id=97bfe604-a0be-4caf-b276-db37b98e3a26
    t.go:99: 2024-07-10 13:15:19.882 [debu]  manager.notifier: dequeued messages  notifier_id=97bfe604-a0be-4caf-b276-db37b98e3a26  count=2
    t.go:99: 2024-07-10 13:15:19.882 [debu]  manager.notifier: message dispatch succeeded  notifier_id=97bfe604-a0be-4caf-b276-db37b98e3a26  msg_id=74f1fd65-d602-42e4-87c9-a83a91abba9f  method=smtp
    t.go:99: 2024-07-10 13:15:19.882 [debu]  manager.notifier: message dispatch succeeded  notifier_id=97bfe604-a0be-4caf-b276-db37b98e3a26  msg_id=4681beed-a763-41a1-badc-b205bfe03ed8  method=smtp
    t.go:99: 2024-07-10 13:15:19.882 [debu]  manager.notifier: dispatch completed  notifier_id=97bfe604-a0be-4caf-b276-db37b98e3a26  count=2
    t.go:99: 2024-07-10 13:15:19.882 [debu]  manager.notifier: check if notifier is paused  notifier_id=97bfe604-a0be-4caf-b276-db37b98e3a26
    t.go:99: 2024-07-10 13:15:19.882 [debu]  manager.notifier: attempting to dequeue messages  notifier_id=97bfe604-a0be-4caf-b276-db37b98e3a26
    t.go:99: 2024-07-10 13:15:19.882 [debu]  manager.notifier: dequeued messages  notifier_id=97bfe604-a0be-4caf-b276-db37b98e3a26  count=2
```

File: coderd/notifications/notifications_test.go [hunk 24]
```diff
@@ -517,6 +517,73 @@ func TestInvalidConfig(t *testing.T) {
 	require.ErrorIs(t, err, notifications.ErrInvalidDispatchTimeout)
 }
 
+func TestNotifierPaused(t *testing.T) {
+	t.Parallel()
+
+	// setup
+	ctx := testutil.Context(t, testutil.WaitLong)
```
Comment (dannykopping) [inline coderd/notifications/notifications_test.go:524]: Nit: btw now we have `setupInMemory` since https://github.com/coder/coder/pull/13863 was merged.

Comment (mtojek) [reply coderd/notifications/notifications_test.go]: Fixed 👍

**Review by dannykopping — APPROVED** [review]
LGTM
